### PR TITLE
AV1: Seeking & Sequence Header rules

### PR DESCRIPTION
**Pull request by mkver**
Here is the promised PR. Sorry for being late, but it took more time than planned.

### CODE REVIEW

File: codec/av1.md [hunk 1]
```diff
@@ -106,14 +104,30 @@ The `Block` timestamp is translated from the __[PresentationTime]__ without the
 
 # Segment Restrictions
 
-Matroska doesn't allow dynamic changes within a codec for the whole `Segment`. The parameters that should not change for a video `Track` are the dimensions and the `CodecPrivate`. 
+AV1 stored in Matroska restricts the allowed variations among the `Sequence Header OBUs` contained in the `CodecPrivate` and also in-band in the `Blocks`. The changes are restricted to the changes that are allowed for the `Sequence Header OBUs` of a `CVS`, i.e. the contents of the `Sequence Header OBUs` MUST be bit-identical each time a `Sequence Header OBU` appears except for the contents of __[operating_parameters_info]__. Furthermore the dimensions of all output frames MUST be equal.
+
+
```
Comment (retokromer) [inline codec/av1.md]: Nit: please only one empty line.

Comment (mkver) [reply codec/av1.md]: Thanks for taking a look, but this is directly in front of a header on the same level as e.g. "Block Data" or "Segment Restrictions" and there are two empty lines in front of these, too. I just imitated the style already in use.

Comment (retokromer) [reply codec/av1.md]: I see, then please ignore my comment. 
(Yet it’s not the case before `# Encryption` just below, which actually gave me the hint ;-)

Comment (mkver) [reply codec/av1.md]: Ok, this is an oversight. Should I just add a new commit or amend the last one (rewriting history is after all frowned upon when one has already made something public and I honestly don't know what will happen with this PR when I replace one of the commits it contains with another one. Is the new/updated commit then automatically part of this PR?)

Comment (retokromer) [reply codec/av1.md]: Oh, in my personal opinion the History will stay exactly the same either way, but possibly my colleagues are more orthodox on this than I am. Let’s wait and see.

File: codec/av1.md [hunk 2]
```diff
@@ -28,6 +25,9 @@ The top Matroska element that contains interleaved audio, video, subtitles as we
 ## Temporal Unit
 All the OBUs that are associated with a time instant. It consists of a `Temporal Delimiter OBU`, and all the OBUs that follow, up to but not including the next `Temporal Delimiter OBU`.
 
+## CVS
+A Coded Video Sequence is a sequence of `Temporal Units` where the contents of all `Sequence Header OBUs` in the bitstream before Matroska encapsulation must be bit-identical except for the contents of __[operating_parameters_info]__.
+
```
Comment (robUx4) [inline codec/av1.md]: Did you just reorder it or change the text ? They are ordered alphabetically. And if that's just reordering then it's not a `Seeking & Sequence Header rules`.

Comment (mkver) [reply codec/av1.md]: I did reorder and change the text. I was not satisfied with the current wording of this. After all, "video frames" could mean: output video frames, (Redundant) Frame (Header) OBUs or Matroska "frames", i.e. (Simple)Blocks. The first of these is not what is meant, the second is also not good because there can be OBUs of a different type in a CVS. And the third possibility simply ignores that CVS are defined in the AV1 specification and that they existed before the AV1 Matroska/Webm mapping. So I thought that I should clarify this. And given that the definition of a CVS is intimately related to what type of AV1 shall be allowed in Matroska I thought that this change belongs right besides the other ones.

As for reordering, I have used `Temporal Units` in the definition of `CVS` so I thought that the latter should be defined after the former.

I think that my second commit should be split from the other ones. Should I force-push an update to my branch (including rewriting history) or do something else?

File: codec/av1.md [hunk 3]
```diff
@@ -87,11 +87,9 @@ The `Redundant Frame Header OBUs` SHOULD not be used.
 
 OBU trailing bits SHOULD be limited to byte alignment and SHOULD not be used for padding.
 
-`Sequence Header OBUs` SHOULD be omitted when they are bit-identical to the one found in `CodecPrivate` and no __[operating_parameters_info]__ exists and the previous `Sequence Header OBUs` in the bistream, if any, was also bit-identical to the one found in `CodecPrivate`. They can be kept when encryption constraints require it.
+A `SimpleBlock` MUST NOT be marked as a Keyframe if it doesn't contain a `Frame OBU`. A `SimpleBlock` MUST NOT be marked as a Keyframe if the first `Frame OBU` doesn't have a __[frame_type]__ of `KEY_FRAME`.
 
-A `SimpleBlock` MUST NOT be marked as a Keyframe if it doesn't contain a `Frame OBU`. A `SimpleBlock` MUST NOT be marked as a Keyframe if the first `Frame OBU` doesn't have a __[frame_type]__ of `KEY_FRAME`. A `SimpleBlock` MUST NOT be marked as a Keyframe if it doesn't contains a `Sequence Header OBU` unless the `Sequence Header OBU` is correctly omitted (see above).
```
[the remark this answers is not in the document]
Comment (mkver) [reply codec/av1.md]: The requirement that a SimpleBlock marked as keyframe or the equivalent Block MUST contain the right `Sequence Header OBU` is still in my proposal; it's in the "A muxer MUST make sure that..." paragraph under the Headline "Seeking and Sequence Header Requirements" where I think it belongs.

File: codec/av1.md [hunk 4]
```diff
@@ -108,12 +106,28 @@ The `Block` timestamp is translated from the __[PresentationTime]__ without the
 
 AV1 stored in Matroska restricts the allowed variations among the `Sequence Header OBUs` contained in the `CodecPrivate` and also in-band in the `Blocks`. The changes are restricted to the changes that are allowed for the `Sequence Header OBUs` of a `CVS`, i.e. the contents of the `Sequence Header OBUs` MUST be bit-identical each time a `Sequence Header OBU` appears except for the contents of __[operating_parameters_info]__. Furthermore the dimensions of all output frames MUST be equal.
 
-The first `Sequence Header OBU` of a `CVS` is stored in the `CodecPrivate` of a `Track`.
 
-If a __[operating_parameters_info]__ exists in a `Sequence Header OBU` then each keyframe `Block` MUST contain a `Sequence Header OBU` before the `Frame Header OBUs`.
+# Seeking and Sequence Header Requirements
+
+`SimpleBlocks` with the `Keyframe` flag set and `Blocks` inside `BlockGroups` that do not contain any `ReferenceBlock` elements constitute the Random Access Points (RAP) of a Matroska `Track`.
```
[the remark this answers is not in the document]
Comment (mkver) [reply codec/av1.md]: A `Block` is always inside a `BlockGroup` and so are any potentially existing `ReferenceBlocks`. It is the `BlockGroups` that don't contain `ReferenceBlocks` that this clause speaks of and not the `Blocks` that do not contain any `ReferenceBlock` (as this is automatically fulfilled).

File: codec/av1.md [hunk 5]
```diff
@@ -108,12 +106,28 @@ The `Block` timestamp is translated from the __[PresentationTime]__ without the
...
 
-The first `Sequence Header OBU` of a `CVS` is stored in the `CodecPrivate` of a `Track`.
 
-If a __[operating_parameters_info]__ exists in a `Sequence Header OBU` then each keyframe `Block` MUST contain a `Sequence Header OBU` before the `Frame Header OBUs`.
+# Seeking and Sequence Header Requirements
+
+`SimpleBlocks` with the `Keyframe` flag set and `Blocks` inside `BlockGroups` that do not contain any `ReferenceBlock` elements constitute the Random Access Points (RAP) of a Matroska `Track`.
+
+Upon random access to a RAP or upon starting playback from the beginning the player/demuxer MUST prepend the `Sequence Header OBU` contained in the `CodecPrivate` to the data of the `Block` where decoding starts after having reverted any potentially applying `ContentCompression`. Afterwards playback proceeds normally including consumption of any `Sequence Header OBUs` that are found in the bitstream.
```
[the remark this answers is not in the document]
Comment (mkver) [reply codec/av1.md]: It's obvious to you and me, but is it really obvious to everyone? Image a track using header removal compression that applies to the data of the (Simple)Blocks. Does really everybody know in which order the prepending of the `Sequence Header OBUs` and the prepending of whatever should be prepended to the data should happen? I am not so sure and in order to avoid confusion I inserted this clause.
I did indeed not think about the scope of the `ContentCompression`, but now that I think about it my wording ("any potentially applying") fits even when taking this into account.

File: codec/av1.md [hunk 6]
```diff
@@ -108,12 +106,28 @@ The `Block` timestamp is translated from the __[PresentationTime]__ without the
...
+
+Seeking to a non-RAP is undefined and discouraged.
+
+A muxer MUST make sure that when using a conformant demuxer/player the correct `Sequence Header OBU` is active both during linear access and also after random access to a RAP. In particular, if a `Sequence Header OBU` whose contents differ from the contents of the `Sequence Header OBU` in the `CodecPrivate` is active during consumption of the first `Frame Header OBU` of a RAP when performing linear access from the beginning, then said `Block` MUST contain said `Sequence Header OBU` in front of the first `Frame Header OBU` so that it is also active when performing random access to said RAP.
```
[the remark this answers is not in the document]
Comment (mkver) [reply codec/av1.md]: The active `Sequence Header OBU` is the `Sequence Header OBU` currently in use, i.e. the one that will be used to decode the frames. If you want me to explain this in more detail than I actually think is necessary I would use the following (inspired from section 7.4.1.2.1 of the H.264 standard):
"A `Sequence Header OBU` is considered active from the point of the bitstream on where it appears and it stays active until a new `Sequence Header OBU` is later encountered in the bitstream. Said new `Sequence Header OBU` is active from this point onwards."

File: codec/av1.md [hunk 7]
```diff
@@ -108,12 +106,28 @@ The `Block` timestamp is translated from the __[PresentationTime]__ without the
...
+
+A muxer MAY omit (strip away and discard) `Sequence Header OBUs` provided the above criteria are still fulfilled afterwards.
+
+*Notes: a) The contents of two `Sequence Header OBUs` of an AV1 bitstream compliant with the requirements for AV1 tracks in Matroska can only differ if __[decoder_model_present_for_this_op]__ is 1 for some operating point for one (hence every) `Sequence Header OBU`. In particular, if __[timing_info_present]__ is set to 0 for any (hence every) `Sequence Header OBU`, all `Sequence Header OBUs` can be omitted from the bitstream.*
```
[the remark this answers is not in the document]
Comment (mkver) [reply codec/av1.md]: The [decoder_model_present_for_this_op[ i ]] values are part of the things that must coincide for all the `Sequence Header OBUs` of a `CVS`. So if we only have (or presume to have) a `CVS` and if for one `Sequence Header OBU` all [decoder_model_present_for_this_op[ i ]] are 0 (for all applicable i), then all [decoder_model_present_for_this_op[ i ]] for every `Sequence Header OBU` in the whole `CVS`. This equivalence is behind the "hence every" in the above statement. This equivalence is important because it enables a muxer to know whether there will be any in-band `Sequence Header OBUs` right at the beginning of muxing (after analyzing the first `Sequence Header OBU`).

One could shorten this to: "If a `CVS` contains a `Sequence Header OBU` for which __[decoder_model_present_for_this_op]__ is 0 for all present operating points, then the contents of all `Sequence Header OBUs` contained in this `CVS` are bit-identical to each other and hence can be omitted. This applies in particular if __[timing_info_present]__ is set to 0 in any `Sequence Header OBU`.

File: codec/av1.md [hunk 8]
```diff
@@ -108,12 +106,28 @@ The `Block` timestamp is translated from the __[PresentationTime]__ without the
...
+
+*b) A `Sequence Header OBU` that fulfills any of these criteria can be omitted without changing compliance to the above criteria:*
```
[the remark this answers is not in the document]
Comment (mkver) [reply codec/av1.md]: The "these criteria" refers to the following three paragraphs; the "above criteria" are the criteria regarding in-band `Sequence Header OBUs` that a muxer must fulfill (i.e. it is the "A muxer MUST make sure" paragraph).

File: codec/av1.md [hunk 9]
```diff
@@ -106,14 +104,30 @@ The `Block` timestamp is translated from the __[PresentationTime]__ without the
...
+
+*Notes: a) The contents of two `Sequence Header OBUs` of an AV1 bitstream compliant with the requirements for AV1 tracks in Matroska can only differ if __[decoder_model_present_for_this_op]__ is 1 for some operating point for one (hence every) `Sequence Header OBU`. In particular, if __[timing_info_present]__ is set to 0 for any (hence every) `Sequence Header OBU`, all `Sequence Header OBUs` can be omitted from the bitstream.*
 
-The first `Sequence Header OBU` of a `CVS` is stored in the `CodecPrivate` of a `Track`. So this AV1 `Track` has the same requirements as the `CVS`.
+*b) A `Sequence Header OBU` that fulfills any of these criteria can be omitted without changing compliance to the above criteria:*
 
-If the __[decoder_model_info_present_flag]__ of this `Sequence Header OBU` is set to 1 then each keyframe `Block` MUST contain a `Sequence Header OBU` before the `Frame Header OBUs`.
+*If, after potentially stripping away all `Temporal Delimiter OBUs`, there are more than one `Sequence Header OBUs` immediately after one another, then all but the last of these `Sequence Header OBUs` can be omitted.*
```
[the remark this answers is not in the document]
Comment (mkver) [reply codec/av1.md]: "Potentially" means that one can check this criterion either before or after stripping the `Temporal Delimiter OBUs` away. After is of course better as this means that more superfluous `Sequence Header OBUs` will be detected as omitable, but implementations are of course free to do is as they see fit. It does not imply that stripping away `Temporal Delimiter OBUs` is optional.

Btw: One can strip even more away before applying this test; i.e. one can discard all `OBU_PADDING` `OBUs`. Notice that the syntax of an `OBU` of type `OBU_METADATA` with [metadata_type] "Unregistered user private" may depend on the `Sequence Header OBU` currently active so that simply ignoring `OBU_METADATA` `OBUs` for this test might not be right.

File: codec/av1.md [hunk 10]
```diff
@@ -106,14 +104,30 @@ The `Block` timestamp is translated from the __[PresentationTime]__ without the
...
 
-Given a `Sequence Header OBU` can be omitted from a `Block` if __[decoder_model_info_present_flag]__ is 0 and it is bit identical to the one found in `CodecPrivate`, when seeking to a keyframe, that omitted `Sequence Header OBU` MUST be added back to the bitstream for compliance with the Random Access Decoding section of the [AV1 Specifiations](#av1-specifications).
+*A `Sequence Header OBU` whose contents differ from the contents of the `Sequence Header OBU` in the `CodecPrivate` can be omitted if there was a preceding `Sequence Header OBU` in the bitstream, if the contents of the the preceding `Sequence Header OBU` were bit-identical to the contents of the current `Sequence Header OBU` and if the current `Sequence Header OBU` is not the first OBU after a `Temporal Delimiter OBU` that starts a temporal unit whose corresponding `Block` is a RAP.*
```
[the remark this answers is not in the document]
Comment (mkver) [reply codec/av1.md]: No, this is not a new rule. This just gives you a criterion which tells you whether it is safe to omit a `Sequence Header OBU` according to the rules above (the rules above means the two paragraphs above the notes that contain what a muxer MUST and what a muxer MAY do).

It is safe to apply these rules in any order; one should then apply the next rule on the output after the last check.

File: codec/av1.md [hunk 11]
```diff
@@ -106,14 +104,30 @@ The `Block` timestamp is translated from the __[PresentationTime]__ without the
...
 
+*A `Sequence Header OBU` whose contents are bit-identical to the contents of the `Sequence Header OBU` in the `CodecPrivate` can be omitted if there was no preceding `Sequence Header OBU` or if there was a preceding `Sequence Header OBU` and the contents of the preceding `Sequence Header OBU` were bit-identical to the contents of the current `Sequence Header OBU`.*
```
[the remark this answers is not in the document]
Comment (mkver) [reply codec/av1.md]: The last paragraph contained a criterion for `Sequence Header OBUs` that differ from the one contained in the `CodecPrivate`, this one is for the `Sequence Header OBUs` that coincide with the one in the `CodecPrivate`. These two criteria are not the same and I would say they are not even "pretty much the same", although there is a certain resemblance between these two (i.e. both criteria contain references to the preceding `Sequence Header OBU`).

File: codec/av1.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (robUx4) [reply codec/av1.md]: We're not going to explain how each Matroska element works (like DisplayWidth/Height) because this is a new document. IMO `ContentCompression` is completely orthogonal to this spec and should not be taken in consideration at all. If that happens it's outside of this spec and transparent.
We also don't define what should be put in `Cues` even though it has an impact on seeking. (but now that I think of it we might, because Switch frames can do good Cue points even though they are not keyframes (AFAIK)).

File: codec/av1.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (robUx4) [reply codec/av1.md]: Yes, that sounds good if we're going to use it.
In the "Seeking to a non-RAP" paragraph I still understand what it means after the word "active". You should rephrase it, maybe using more sentences.

File: codec/av1.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (robUx4) [reply codec/av1.md]: It can be simplified by using the "active Sequence Header OBU" (which might have a definition in the Terms section).

**Comment by jamrial** [comment]
> Refining the Sequence Header OBU handling is needed. I will put it on hold for now as it may change completely if we decide we don't want it at all in the CodecPrivate. (and instead use a custom structure as in ISOBMFF with just the fields needed to init the playback).

The ISOBMFF spec was just updated [1] to include basic stream information as well as optional OBUs in av1C, guaranteeing the presence of enough init data before a block has to be parsed, so it would be great for the sake of compatibility and simplicity in format support if Matroska could include the contents of AV1CodecConfigurationRecordv1 verbatim inside CodecPrivate, much like it's done for h264 and hevc, instead of the currently defined raw OBUs.

[1] https://github.com/AOMediaCodec/av1-isobmff/pull/64

**Comment by jamrial** [comment]
Furthermore, unlike you defined in this spec, for ISOBMFF they don't allow keyframes to omit the Sequence Header OBU if it's equal to the one in CodecPrivate/configOBUs. See https://github.com/AOMediaCodec/av1-isobmff/issues/67

This will result in an unnecessary added complexity when remuxing mkv -> mp4 if the Matroska file doesn't have the SH in key frames, so it would also be convenient to match their behavior and remove said flexibility, keeping the bitstream within keyframes as fully compliant RAPs.

**Comment by robUx4** [comment]
I totally agree, we'll copy the struct as it is in ISOBMFF. And not allow removing the OBUs from the bitstream. We'll lose a bit of space but we can always use a new "Header Compression" mode that would only apply to keyframes and achieve the same goal. That would be for a future Matroska/WebM version.

**Comment by robUx4** [comment]
Since there is now a structure at the start of the `CodecPrivate` and the `Sequence Header OBU` are not stripped from the stream anymore, can we close this PR ?

**Comment by mbunkus** [comment]
As far as I can see this PR not only contains the seeking rules but the whole AV1-in-Matroska mapping, correct? I further assume that It's your intention to continue working on the mapping in its own branch and only to discard the seeking rules that were the main point of this PR? In that case I concur, the PR can be closed.

**Comment by robUx4** [comment]
Yes, this PR is on the av1-mappin branch which is still being worked on.

**Comment by mbunkus** [comment]
Alright, so let's close it.
